src/util/RatelimitBucket.hpp

```diff
@@ -25,16 +41,32 @@ class RatelimitBucket : public QObject
      **/
     const int cooldown_;
 
-    std::function<void(QString)> callback_;
-    QList<QString> queue_;
+    std::function<void(T)> callback_;
+    QList<T> queue_;
 
     /**
      * @brief Run the callback on one entry in the queue.
      *
      * This will start a timer that runs after cooldown_ milliseconds that
      * gives back one "token" to the bucket and calls handleOne again.
      **/
-    void handleOne();
+    void handleOne()
+    {
+        if (this->queue_.isEmpty())
+        {
+            return;
+        }
+
+        auto item = this->queue_.takeFirst();
+
+        this->budget_--;
+        this->callback_(item);
+
+        QTimer::singleShot(cooldown_, this, [this] {
+            this->budget_++;
```

Comment:
consider making budget atomic if `submit` could be called separately from the qt event loop

Reply:
For now, I opted to make the minimal required changes to `RatelimitBucket` but I will consider this